### DISCUSSION TITLE
Callback function on termination condition or human in the loop

### DISCUSSION THREAD

**Discussion by chandan867** [microsoft/autogen#4033] [Ideas]
After spending some time with Autogen, I felt need for couple of improvements:

1. Callback Function: Surprised there's no built-in way to trigger functions when terminal conditions are met—like if a 'reviewer' agent gives a rating of 8 or higher. Right now, custom implementation seems to be the only option.

2. Human-in-the-Loop Integration: For user prompts, relying on the terminal feels limiting. I'd like users to interact via WhatsApp or Slack instead.

**Comment by ekzhu** [comment]
@chandan867  is your experience based on v0.2 or v0.4 API? We are working on the new v0.4 API, which aims to fix these concerns. My response here is based on v0.4.

Regarding call back function, in AgentChat, once the task is terminated, a task result object is returned to the application. You can then perform necessary operation based on the result. In the Core API, since you define your agent classes, it's up to you to add call backs (e.g., as an event handler)

Regarding human-in-the-loop integration, you are spot on. The Core API is going to be the foundation for making it happen. We are looking to improve the documentation and samples. See: #4001 and https://github.com/microsoft/autogen/discussions/3915. The AgentChat API will also support human-in-the-loop interface and hooks. Though, we would appreciate help in building the connectors to apps like Slack and WhatsApp.
